src/connectivity.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * Connectivity Checker to perform basic connectivity checks with BrowserStack
+ * components, i.e. Hub and Rails.
+ * It performs checks with/without proxy, over HTTP(S).
+ * NOTE : HTTPS with Proxy is not implemented since the tool is anyway capturing HTTP
+ * traffic only. Shall be taken up later.
+ */
+
+var http = require('http');
+var url = require('url');
+var constants = require('../config/constants');
+var RdGlobalConfig = constants.RdGlobalConfig;
+var Utils = require('./utils');
+var https = require('https');
+
+/**
+ * Fires the requests to perform connectivity checks
+ * @param {Object} requestOptions 
+ * @param {'http'|'https'} requestType 
+ * @param {String} description 
+ * @param {Array<Number>} successCodes 
+ * @param {Function} callback 
+ */
+var fireRequest = function (requestOptions, requestType, description, successCodes, callback) {
+  var httpOrHttps = requestType === 'http' ? http : https;
+  var responseData = {
+    data: [],
+    statusCode: null,
+    errorMessage: null,
+    description: description,
+    result: 'Failed'
+  };
+
+  var request = httpOrHttps.request(requestOptions, function (response) {
+    responseData.statusCode = response.statusCode;
+    if (successCodes.indexOf(response.statusCode) !== -1) {
+      responseData.result = "Passed";
+    }
+
+    response.on('data', function (chunk) {
+      responseData.data.push(chunk);
+    });
+
+    response.on('end', function () {
+      responseData.data = Buffer.concat(responseData.data).toString();
+      callback(responseData);
+    });
+
+  });
+
+  request.on('error', function (err) {
+    responseData.errorMessage = err.toString();
+    callback(responseData);
+  });
+
+  request.setTimeout(constants.CONNECTIVITY_REQ_TIMEOUT, function () {
```

Comment:
Can `CONNECTIVITY_REQ_TIMEOUT` be made configurable?

Reply:
It can be made configurable. But its not a major value as such. The current value of 20 seconds is more than enough for those endpoints since they just reach out to components with GET request, i.e. check status at Hub or check for redirection response from Rails. Do you think the user will have to increase this value as well? Since, if the user is not getting any response in 20 seconds, thats a red flag in itself at their end.

Reply:
Resolved F2F